### PR TITLE
Swap [ui part 1]

### PR DESCRIPTION
**Pull request by grod220**
Closes: https://github.com/penumbra-zone/web/issues/390

The first part of the multi-part PR series in doing swaps. Does a few things:

- Implements new swap asset selector components
- Refactor swap slice
- Implement transaction wasm planning for swaps

### CODE REVIEW

File: apps/webapp/src/components/header/navbar.tsx [hunk 1]
```diff
@@ -15,8 +15,7 @@ export const Navbar = () => {
             to={link.href}
             className={cn(
               'font-bold py-[10px] select-none',
-              link.subLinks &&
-                link.subLinks.includes(pathname) &&
+              (link.href === pathname || link.subLinks?.includes(pathname)) &&
```
Comment (grod220) [inline apps/webapp/src/components/header/navbar.tsx:18]: Felt a bit annoying that a path had to add itself to its own set of sublinks

File: apps/webapp/src/components/swap/asset-out-box.tsx
Comment (grod220) [inline apps/webapp/src/components/swap/asset-out-box.tsx:1]: Screenshot 2024-01-29 at 4 01 16 PM

File: apps/webapp/src/components/swap/asset-out-selector.tsx
Comment (grod220) [inline apps/webapp/src/components/swap/asset-out-selector.tsx:1]: Screenshot 2024-01-29 at 4 01 21 PM

File: packages/storage/src/indexed-db/index.ts [hunk 2]
```diff
@@ -223,46 +223,6 @@ export class IndexedDb implements IndexedDbInterface {
     }
   }
 
-  private addSctUpdates(txs: IbdUpdates, sctUpdates: ScanBlockResult['sctUpdates']): void {
```
Comment (grod220) [inline packages/storage/src/indexed-db/index.ts:226]: A lint in webstorm re-orders these fyi (alphabetically?)

File: apps/webapp/src/components/shared/select-token-modal.tsx [hunk 3]
```diff
@@ -11,7 +11,7 @@ import {
 import { fromBaseUnitAmount } from '@penumbra-zone/types';
 import { cn } from '@penumbra-zone/ui/lib/utils';
 import { AccountBalance } from '../../fetchers/balances';
-import { AssetIcon } from './asset-icon';
+import { AssetIcon } from './asset-icon.tsx';
```
Comment (jessepinho) [inline apps/webapp/src/components/shared/select-token-modal.tsx:14]: Is this intentional?

I've noticed that, throughout the web monorepo, we're importing from `some-file.tsx` in some places, and just `some-file` in others. Any idea why that is? Feels like a mistake.

File: packages/constants/assets.ts [hunk 4]
```diff
@@ -1,165 +1,148 @@
...
 };
 
-// TODO: This should only have things like icon or other fields not available in assets proto
-export const assets: Asset[] = [
+export interface Asset {
+  metadata: DenomMetadata;
+  icon?: string;
+}
+
+export const AssetRegistry: Asset[] = [
```
Comment (jessepinho) [inline packages/constants/assets.ts:25]: Not at all a blocker, but what do you think of making this an object, keyed by e.g., display name? Then you could just look up metadata on the object by name, rather than iterating over an array and memoizing the result like you're doing [here](https://github.com/penumbra-zone/web/pull/419/files#diff-4ec70347e05c7bcd69a141c28ebc60dfbf558278548fe0b51ce29b002ff07188R7).

File: apps/webapp/src/state/swap.ts [hunk 5]
```diff
@@ -1,103 +1,105 @@
...
-  pay: SwapAssetInfo;
-  receive: SwapAssetInfo;
-  validationErrors: SwapValidationFields;
-  setAmount: (type: SwapInputs) => (amount: string) => void;
-  setAsset: (type: SwapInputs) => (asset: AssetId) => void;
-  replaceAsset: () => void;
-  setAssetBalance: (amount: number) => void;
+  assetIn: Selection | undefined;
+  setAssetIn: (selection: Selection) => void;
+  amount: string;
```
Comment (jessepinho) [inline apps/webapp/src/state/swap.ts:20]: Shouldn't this be a `number`? (and get converted in the `InputToken`'s [`onChange` handler](https://github.com/penumbra-zone/web/pull/419/files#diff-0a9c5fc7496ea1a47f7fef2e351c40f80cdc243873cbf333c5412ddac5ee1f81R30-R33))?

File: apps/webapp/src/components/swap/swap-form.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      <InputToken
+        label='Amount to swap'
+        placeholder='Enter an amount'
+        className='mb-1'
+        selection={assetIn}
+        setSelection={setAssetIn}
+        value={amount}
+        onChange={e => {
+          if (Number(e.target.value) < 0) return;
+          setAmount(e.target.value);
```
Comment (jessepinho) [inline apps/webapp/src/components/swap/swap-form.tsx:32]: See my note in the Zustand state — shouldn't this be converted to a number first?

File: apps/webapp/src/components/swap/asset-out-box.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+export const AssetOutBox = ({ balances }: AssetOutBoxProps) => {
+  const { assetOut, setAssetOut } = useStore(swapSelector);
+
+  const aggregatedBalances = balances
+    .flatMap(b => b.balances)
+    .reduce<AssetBalance[]>(groupByAsset, []);
+
+  const balanceOfDenom = aggregatedBalances.find(b =>
+    b.assetId.equals(assetOut?.metadata.penumbraAssetId),
+  );
```
Comment (jessepinho) [inline apps/webapp/src/components/swap/asset-out-box.tsx:21]: I'm personally a fan of doing as little logic in the template as possible, so I'd prefer adding another const below this one, something like this:
```suggestion
  );
  const formattedBalanceOfDenom = balanceOfDenom
              ? fromBaseUnitAmount(balanceOfDenom.amount, balanceOfDenom.denom.exponent).toFormat()
              : '0';
```

Then, you can just render `{formattedBalanceOfDenom}` in the markup below.

(Not a blocker though — just a nitpick.)

File: apps/webapp/src/state/send.ts [hunk 8]
```diff
@@ -97,7 +97,7 @@ const assembleRequest = async ({ amount, recipient, selection, memo }: SendSlice
     ],
     source: new AddressIndex({ account: selection.accountIndex }),
     memo: new MemoPlaintext({
-      returnAddress: await getAddressByIndex(selection.accountIndex),
+      returnAddress: await getEphemeralAddress(selection.accountIndex),
```
Comment (jessepinho) [inline apps/webapp/src/state/send.ts:100]: Wait, why this change? We always want sends to be from an ephemeral address?

File: apps/webapp/src/state/swap.ts [hunk 9]
```diff
@@ -1,103 +1,105 @@
...
+          state.send.txInProgress = false;
+        });
+      } catch (e) {
+        set(state => {
+          state.send.txInProgress = false;
+        });
+        dismiss();
+        toastFn(errorTxToast(e));
+        throw e;
+      }
```
Comment (jessepinho) [inline apps/webapp/src/state/swap.ts:74]: Up to you, but some of this could be pulled into a `finally`:

```suggestion
        toastFn(successTxToast(txHash));

        // Reset form
        set(state => {
          state.send.amount = '';
        });
      } catch (e) {
        toastFn(errorTxToast(e));
        throw e;
      } finally {
        set(state => {
          state.send.txInProgress = false;
        });
        dismiss();
      }
```

File: apps/webapp/src/state/swap.ts [hunk 10]
```diff
@@ -1,103 +1,105 @@
...
+
+  return new TransactionPlannerRequest({
+    swaps: [
+      {
+        targetAsset: assetOut.metadata.penumbraAssetId,
+        value: {
+          amount: toBaseUnit(BigNumber(amount), assetIn.asset.denom.exponent),
+          assetId: assetIn.asset.assetId,
+        },
+        claimAddress: await getEphemeralAddress(assetIn.accountIndex),
```
Comment (jessepinho) [inline apps/webapp/src/state/swap.ts:91]: Why an ephemeral address? Is that just a default we want to use? (Seems unnecessary in a swap though, since no one else can read anything in the transaction, right?)

File: packages/router/src/grpc/view-protocol-server/transaction-planner.ts [hunk 11]
```diff
@@ -28,10 +48,27 @@ export const transactionPlanner: Impl['transactionPlanner'] = async (req, ctx) =
...
+    if (!value || !targetAsset || !fee || !claimAddress)
+      throw new Error('no value, targetAsset, fee or claimAddress in swap');
+
+    const intoDenomMetadata = await getDenomMetadata(indexedDb, targetAsset, querier);
+    planner.swap(value, intoDenomMetadata, fee, claimAddress);
+  }
+
+  for (const { swapCommitment } of req.swapClaims) {
+    if (!swapCommitment) throw new Error('no swapCommitment in swapClaim');
+    await planner.swapClaim(swapCommitment);
```
Comment (jessepinho) [inline packages/router/src/grpc/view-protocol-server/transaction-planner.ts:66]: Interesting... so, we can plan the swap and the swap claim at the same time, even though they'll be in separate transactions/blocks?

File: apps/webapp/src/components/swap/swap-form.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import { Button } from '@penumbra-zone/ui';
+import { useToast } from '@penumbra-zone/ui/components/ui/use-toast';
+import InputToken from '../shared/input-token.tsx';
+import { useLoaderData } from 'react-router-dom';
+import { AccountBalance } from '../../fetchers/balances.ts';
+import { useStore } from '../../state';
+import { swapSelector } from '../../state/swap.ts';
+import { AssetOutBox } from './asset-out-box.tsx';
+
+export const SwapForm = () => {
```
Comment (jessepinho) [inline apps/webapp/src/components/swap/swap-form.tsx:10]: I can't actually navigate to this UI in my local git repo to test it, since I'm getting the "missing trailer" issue we're discussing in Discord. But the code looks fine!

**Review by jessepinho — COMMENTED** [review]
LGTM, with some suggestions/nitpicks that I'll leave up to you. I also am still unclear on how swaps + swap claims will be handled in the UI, since they're separate transactions/blocks.

**Review by jessepinho — APPROVED** [review]
whoops, forgot to approve

File: apps/webapp/src/components/swap/asset-out-box.tsx (hunk 7 above)
Comment (hdevalence) [reply apps/webapp/src/components/swap/asset-out-box.tsx:21]: why is this component manually formatting the balances, instead of having that logic only in one place, in a component rendering a `ValueView` (the same way we only format addresses once in a component rendering an `AddressView`)?

File: apps/webapp/src/state/send.ts (hunk 8 above)
Comment (hdevalence) [reply apps/webapp/src/state/send.ts:100]: i'm guessing this was a rebase or something, but we never want to use ephemeral addresses here

File: apps/webapp/src/state/swap.ts (hunk 10 above)
Comment (hdevalence) [reply apps/webapp/src/state/swap.ts:91]: we want to avoid using one-time addresses anywhere except ibc deposits, where they're absolutely necessary, this should indeed be using the default address for the account

File: packages/storage/src/indexed-db/index.ts (hunk 2 above)
Comment (turbocrime) [reply packages/storage/src/indexed-db/index.ts:226]: can we avoid lint/format changes that aren't defined in ci by eslint

File: packages/router/src/grpc/view-protocol-server/transaction-planner.ts (hunk 11 above)
Comment (hdevalence) [reply packages/router/src/grpc/view-protocol-server/transaction-planner.ts:66]: i don't know about the surrounding context, but in general, planning the swap claim requires knowledge of the `BatchSwapOutputData` for the block the swap was submitted in, which arrives in the `CompactBlock` we use to detect when the swap was committed on-chain. so the swap claim can't be planned at the same time as the swap but it can be planned as soon as the swap is detected

Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/transaction-planner.ts:66]: this would only attempt to plan a swap and swapClaim at the same time if the caller requests that a swap and swapClaim be planned at the same time by including both in the TransactionPlannerRequest.

the caller probably does not have the information to make such a request, unless they can independently plan and build transactions, and write to idb. but it's possible by pb spec.

File: packages/constants/assets.ts (hunk 4 above)
Comment (turbocrime) [reply packages/constants/assets.ts:25]: note there are about a thousand assets that will eventually be in this table

File: apps/webapp/src/components/swap/asset-out-box.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        <p className='text-sm font-bold md:text-base'>Swap into</p>
+      </div>
+      <div className='flex items-center justify-between gap-4'>
+        <Input
+          variant='transparent'
+          type='number'
+          className='font-bold leading-10 md:h-8 md:w-[calc(100%-80px)] md:text-xl  xl:h-10 xl:w-[calc(100%-160px)] xl:text-3xl'
+          // TODO: estimate actual swap out amount
+          value=''
+        />
```
Comment (jessepinho) [inline apps/webapp/src/components/swap/asset-out-box.tsx:35]: Ah, just noticed this — so, this will be done in a follow-up PR?

File: packages/constants/assets.ts [hunk 14]
```diff
@@ -1,165 +1,148 @@
...
+  unbondingToken: new RegExp('^unbonding_'),
+  votingReceipt: new RegExp('^voted_on_'),
 };
 
-// TODO: This should only have things like icon or other fields not available in assets proto
-export const assets: Asset[] = [
+export interface Asset {
+  metadata: DenomMetadata;
+  icon?: string;
+}
```
Comment (turbocrime) [inline packages/constants/assets.ts:23]: this should be renamed so it is never confused with the pb Asset type that is used in many places

File: apps/webapp/src/components/swap/layout.tsx [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    <div className='relative mx-auto grid gap-6 md:grid-cols-2 md:gap-4 xl:max-w-[1276px] xl:grid-cols-3 xl:gap-5'>
+      <div className='hidden xl:order-1 xl:block' />
+      <Card gradient className='order-2 row-span-2 flex-1 p-5 md:order-1 md:p-4 xl:p-5'>
+        <SwapForm />
+      </Card>
+      <EduInfoCard
+        className='order-1 md:order-2'
+        src='./receive-gradient.svg'
+        label='Swap me'
+        content={EduPanel.TEMP_FILLER}
```
Comment (jessepinho) [inline apps/webapp/src/components/swap/layout.tsx:42]: https://film.jessepinho.com/lorem-ipsum

😏

File: packages/router/src/grpc/view-protocol-server/transaction-planner.ts (hunk 11 above)
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/transaction-planner.ts:66]: i had a comment here about why it isn't possible to make such a request but then accidentally deleted it just now, lol. but

1. it's valid pb to make such a request
2. caller doesn't have the information to do so unless they can plan/build independently
3. such a request would likely fail in wasm, which queries idb for that swap commitment, which should not be present yet

File: packages/constants/assets.ts (hunk 14 above)
Comment (hdevalence) [reply packages/constants/assets.ts:23]: I don't have context here but why do we want this at all? shouldn't the icon be part of the `DenomMetadata`?

**Comment by hdevalence** [comment]
I think it could be good to minimize the amount of further changes we do to asset info in this PR and work on an asset registry as a separate effort, since it's something of a project in its own right.

One thing I noticed by looking at this PR is that the existing web code has a lot of ad hoc types for asset information. I think it would be good to just use `DenomMetadata` everywhere, and ensure it's fit for that purpose. We shouldn't be ending up in a place where the web code has to define a type that's "`DenomMetadata` plus an icon", the `DenomMetadata` should just have that field. It's a type we made up, after all.

File: packages/types/src/transaction/perspective.ts [hunk 16]
```diff
@@ -43,6 +47,20 @@ export const viewActionFromEmptyPerspective = (action: Action): ActionView | und
...
+          value: new SwapView({
+            swapView: {
+              case: 'opaque',
+              value: new SwapView_Opaque({
+                swap: action.action.value,
+              }),
+            },
+          }),
+        },
+      });
```
Comment (turbocrime) [inline packages/types/src/transaction/perspective.ts:63]: i guess this matches other sections of this file, but i'd like to point out you don't have to nest constructors like this - it's fine to just use the top-level type

```suggestion
    case 'swap':
      return new ActionView({
        actionView: {
          case: 'swap',
          value: {
            swapView: {
              case: 'opaque',
              value: {
                swap: action.action.value,
              },
            },
          },
        },
      });
```

**Review by turbocrime — COMMENTED** [review]
this needs some rebase, notably i went ahead and killed those weird asset types, and the new proto specs change methods to streaming, which all definitely has downstream effects on you. but otherwise nothing stands out and im sleepy
